firebase_admin/messaging.py

```diff
@@ -299,6 +446,26 @@ def make_topic_management_request(self, tokens, topic, operation):
         else:
             return TopicManagementResponse(resp)
 
+    def _message_data(self, message, dry_run):
+        data = {'message': _MessagingService.encode_message(message)}
+        if dry_run:
+            data['validate_only'] = True
+        return data
+
+    def _postproc(self, resp, body):
+        """Handle response from batch API request."""
+        if resp.status is not 200:
+            data = {}
+            try:
+                parsed_body = json.loads(body.decode())
+                if isinstance(parsed_body, dict):
+                    data = parsed_body
+            except ValueError:
+                pass
+            code, msg = _MessagingService._parse_fcm_error(data, body, resp.status)
+            raise ApiCallError(code, msg)
```

Comment:
I think I understand why this is not working properly. I found the following code in the `googleapiclient`:

```
     try:
        if resp.status >= 300:
          raise HttpError(resp, content, uri=request.uri)
        response = request.postproc(resp, content)
      except HttpError as e:
        exception = e
```

So we need to raise a `googleapiclient.errors.HttpError` from our `postproc`. Then we can get the content from it, and turn it into an `ApiCallError` in our callback.

Reply:
This is my confusion - if we get a >= 300 from whatever our API call returns, we don't hit our `postproc` function. We catch the `googleapiclient.errors.HttpError` https://github.com/firebase/firebase-admin-python/pull/283/files#diff-3101917f5d55cf93592ea47ca59a7015R409

Our `postproc` appears to only be called when we get < 300 status code. Which was my question earlier about "when would we get a non-200 and non-error response code from FCM". It feels like we can just throw a generic error in our `postproc` if it's non-200, right? Other cases should already be handled?

Reply:
FCM mainly responds with 200 and 4xx/5xx responses. In case of 4xx/5xx the above code will raise an `HTTPError`, which we can catch in our callback, and convert into an `ApiCallError`. I believe you're already doing this. 

So that brings us to `postproc`. You're right that `postproc` will only execute for 2xx responses. AFAIK FCM only returns 200 in this range. So we can relax our logic a little bit as follows:

```
def _postproc(self, resp, body):
        """Handle response from batch API request."""
        # This only gets called for 2xx responses.
        return json.loads(body.decode())
```